central/role/service/service_impl.go

```diff
@@ -365,3 +488,212 @@ func effectiveAccessScopeForSimpleAccessScope(scopeRules *storage.SimpleAccessSc
 
 	return response, nil
 }
+
+const (
+	hasFullAccess    = true
+	hasPartialAccess = false
+)
+
+func listReadPermissions(
+	requestedPermissions []string,
+	scope permissions.ResourceScope,
+) []permissions.ResourceWithAccess {
+	readPermissions := resources.AllResourcesViewPermissions()
+	indexedScopeReadPermissions := make(map[string]permissions.ResourceWithAccess, 0)
+	scopeReadPermissions := make([]permissions.ResourceWithAccess, 0, len(readPermissions))
+	for _, permission := range readPermissions {
+		if permission.Resource.GetScope() >= scope {
+			scopeReadPermissions = append(scopeReadPermissions, permission)
+			indexedScopeReadPermissions[permission.Resource.String()] = permission
+		}
+	}
+	if len(requestedPermissions) == 0 {
+		return scopeReadPermissions
+	}
+	scopeRequestedReadPermissions := make([]permissions.ResourceWithAccess, 0, len(scopeReadPermissions))
+	deduplicatedRequestedPermissions := set.NewStringSet()
+	for _, permission := range requestedPermissions {
+		deduplicatedRequestedPermissions.Add(permission)
+	}
+	for _, permission := range deduplicatedRequestedPermissions.AsSlice() {
```

Comment:
You could reuse our `sliceutils` package here as such:
```suggestion
	for _, permission := range sliceutils.Unique(requestedPermissions) {
```